In [9]:
from functools import reduce
import pandas as pd
import json

In [10]:
# Get info such as number of pdb, units...
f = open('data/entries.json')
pdbs_list = pd.read_csv('data/input.txt', sep=' ')

data = json.load(f)
data = pd.DataFrame.from_records(data)
data = data.loc[data['class'] != ""]
data = pd.merge(data, pdbs_list, on=['pdb_id', 'pdb_chain'], how='right')

In [11]:
 # pdb chains
pdbs_top = data.groupby(['repeatsdb_id','class_topology'])['class_topology'].size().reset_index(name='PDB chains')
pdbs_top = pdbs_top.groupby(['class_topology']).size().reset_index(name='PDB chains')
pdbs_top = pdbs_top.append(pdbs_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
pdbs_cl = data.groupby(['repeatsdb_id','class'])['class'].size().reset_index(name='PDB chains')
pdbs_cl = pdbs_cl.groupby(['class']).size().reset_index(name='PDB chains')
pdbs_cl = pdbs_cl.append(pdbs_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')

/tmp/ipykernel_1480133/186686533.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  pdbs_top = pdbs_top.append(pdbs_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
/tmp/ipykernel_1480133/186686533.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  pdbs_cl = pdbs_cl.append(pdbs_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')


In [16]:
 # regions
regions_top = data.groupby(['region_id','class_topology'])['class_topology'].size().reset_index(name='Regions')
regions_top = regions_top.groupby(['class_topology']).size().reset_index(name='Regions')
regions_top = regions_top.append(regions_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
regions_cl = data.groupby(['region_id','class'])['class'].size().reset_index(name='Regions')
regions_cl = regions_cl.groupby(['class']).size().reset_index(name='Regions')
regions_cl = regions_cl.append(regions_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')

   class  Regions
0      2        1
1      3     1520
2      4     2036
3      5      123
4  Total     3680


/tmp/ipykernel_1480133/3404434773.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  regions_top = regions_top.append(regions_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
/tmp/ipykernel_1480133/3404434773.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  regions_cl = regions_cl.append(regions_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')


In [13]:
 # units
units = data.loc[data['type'] == 'unit']
units_top = units.groupby(['class_topology'])['class_topology'].size().reset_index(name='Units')
units_top = units_top.append(units_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
units_cl = units.groupby(['class'])['class'].size().reset_index(name='Units')
units_cl = units_cl.append(units_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')

/tmp/ipykernel_1480133/3976614780.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  units_top = units_top.append(units_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
/tmp/ipykernel_1480133/3976614780.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  units_cl = units_cl.append(units_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')


In [20]:
 # avg_units_len
avg_units_top = units.groupby(['class_topology'])['region_average_unit_length'].mean().round(2).reset_index(name='Unit len')
avg_units_top = avg_units_top.append(avg_units_top.mean().round(2), ignore_index=True).fillna('Total')
avg_units_cl = units.groupby(['class'])['region_average_unit_length'].mean().round(2).reset_index(name='Unit len')
avg_units_cl = avg_units_cl.append({'class': 'Total', 'Unit len': avg_units_cl['Unit len'].mean()}, ignore_index=True)

   class  Unit len
0      2     27.00
1      3     33.58
2      4     39.50
3      5     75.00
4  Total     43.77


/tmp/ipykernel_1480133/2089663907.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  avg_units_top = avg_units_top.append(avg_units_top.mean().round(2), ignore_index=True).fillna('Total')
/tmp/ipykernel_1480133/2089663907.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  avg_units_top = avg_units_top.append(avg_units_top.mean().round(2), ignore_index=True).fillna('Total')
/tmp/ipykernel_1480133/2089663907.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  avg_units_cl = avg_units_cl.append({'class': 'Total', 'Unit len': avg_units_cl['Unit len'].mean()}, ignore_index=True)


In [33]:
# std
std_units_top = units.groupby(['class_topology'])['region_average_unit_length'].std().round(2).reset_index(
    name='Unit len STD')
std_units_top = std_units_top.append({'class_topology': 'Total', 'Unit len STD': std_units_top['Unit len STD'].mean()}, ignore_index=True)
print(std_units_top)
std_units_cl = units.groupby(['class'])['region_average_unit_length'].std().round(2).reset_index(
    name='Unit len STD')
std_units_cl = std_units_cl.append({'class': 'Total', 'Unit len STD': std_units_cl['Unit len STD'].mean()}, ignore_index=True).fillna('Total')


   class_topology  Unit len STD
0             2.2         0.000
1             3.1         5.390
2             3.2         2.020
3             3.3         5.010
4             3.4         3.530
5             3.6         6.610
6             4.1         3.970
7            4.10         2.940
8             4.2         9.720
9             4.3         4.790
10            4.4         6.200
11            4.5         1.610
12            4.6         8.260
13            4.7         0.000
14            4.9         2.490
15            5.1        20.990
16            5.2        25.790
17            5.3        13.920
18            5.4        20.110
19            5.5        10.890
20          Total         7.712


/tmp/ipykernel_1480133/1260490437.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  std_units_top = std_units_top.append({'class_topology': 'Total', 'Unit len STD': std_units_top['Unit len STD'].mean()}, ignore_index=True)
/tmp/ipykernel_1480133/1260490437.py:8: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  std_units_cl = std_units_cl.append({'class': 'Total', 'Unit len STD': std_units_cl['Unit len STD'].mean()}, ignore_index=True).fillna('Total')


In [26]:
# uniprot
uniprot = pd.read_csv('data/pdb_chain_uniprot.tsv',  sep='\t', header=1, index_col='SP_PRIMARY').reset_index()
uniprot['repeatsdb_id']=uniprot['PDB'] + uniprot['CHAIN']
up = uniprot[['SP_PRIMARY', 'repeatsdb_id']]

uniprot_pdb = data.merge(up, how='left', on='repeatsdb_id')
uniprot_top = uniprot_pdb.groupby(['SP_PRIMARY','class_topology'])['class_topology'].size().reset_index(name='UniProt')
uniprot_top = uniprot_top.groupby(['class_topology'])['class_topology'].size().reset_index(name='UniProt')
uniprot_top = uniprot_top.append(uniprot_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
uniprot_cl = uniprot_pdb.groupby(['SP_PRIMARY', 'class'])['class'].size().reset_index(name='UniProt')
uniprot_cl = uniprot_cl.groupby(['class'])['class'].size().reset_index(name='UniProt')
uniprot_cl = uniprot_cl.append(uniprot_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')

/tmp/ipykernel_1480133/1853083527.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  uniprot_top = uniprot_top.append(uniprot_top.sum(numeric_only=True), ignore_index=True).fillna('Total')
/tmp/ipykernel_1480133/1853083527.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  uniprot_cl = uniprot_cl.append(uniprot_cl.sum(numeric_only=True), ignore_index=True).fillna('Total')


In [27]:
class_table = reduce(lambda x, y: pd.merge(x, y, on='class', how='outer'), [pdbs_cl, regions_cl, units_cl, avg_units_cl, std_units_cl, uniprot_cl])
class_table.to_csv("tables/classes_count.csv",index=False)

   class  PDB chains  Regions  Units  Unit len  Unit len STD  UniProt
0      2           1        1      3     27.00         0.000        1
1      3        1462     1520  12174     33.58         9.110      615
2      4        1895     2036  13523     39.50        11.830      927
3      5         123      123    494     75.00        33.630       81
4  Total        3481     3680  26194     43.77         7.712     1624


Table 1. Class overview of the curated RepeatsDB
Total number of PDB chains, regions etc. Multiple PDB chains can map to the same UniProt protein. Unit len AVG is calculated as the average of means of  PDB chain unit lengths in a specific class.

A PDB can contain units of multiple classes, e.g. class 3 and 4. For this reason even if the total number of unique PDBs is 3840, in this table we have a total of 3480 (each class is counted singularly).

In [ ]:
topology_table = reduce(lambda x, y: pd.merge(x, y, on='class_topology', how='outer'), [pdbs_top, regions_top, units_top, avg_units_top, std_units_top, uniprot_top])
topology_table.to_csv("tables/topologies_count.csv",index=False)

Table 2. Topology overview of the curated RepeatsDB
Total number of PDB chains, regions etc.
A PDB can contain units of multiple topologies, e.g. class 3.3 and 3.6. For this reason even if the total number of unique PDBs is 3840, and total classes are 3480, in this table we have a total of 3533 (each topology is counted singularly).
